src/Handler/UnionHandler.php

```diff
@@ -0,0 +1,134 @@
+
+final class UnionHandler implements SubscribingHandlerInterface
+{
+    private static $aliases = ['boolean' => 'bool', 'integer' => 'int', 'double' => 'float'];
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribingMethods()
+    {
+        $methods = [];
+        $formats = ['json', 'xml'];
+
+        foreach ($formats as $format) {
+            $methods[] = [
+                'type' => 'union',
+                'format' => $format,
+                'direction' => GraphNavigatorInterface::DIRECTION_DESERIALIZATION,
+                'method' => 'deserializeUnion',
+            ];
+            $methods[] = [
+                'type' => 'union',
+                'format' => $format,
+                'direction' => GraphNavigatorInterface::DIRECTION_SERIALIZATION,
+                'method' => 'serializeUnion',
+            ];
+        }
+
+        return $methods;
+    }
+
+    public function serializeUnion(
+        SerializationVisitorInterface $visitor,
+        mixed $data,
+        array $type,
+        SerializationContext $context
+    ) {
+        return $this->matchSimpleType($data, $type, $context);
+    }
+
+    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context)
+    {
+        if ($data instanceof \SimpleXMLElement) {
+            throw new RuntimeException('XML deserialisation into union types is not supported yet.');
+        }
+
+        return $this->matchSimpleType($data, $type, $context);
+    }
+
+    private function matchSimpleType(mixed $data, array $type, Context $context)
+    {
+        $dataType = $this->determineType($data, $type, $context->getFormat());
+        $alternativeName = null;
+
+        if (isset(static::$aliases[$dataType])) {
+            $alternativeName = static::$aliases[$dataType];
+        }
+
+        foreach ($type['params'] as $possibleType) {
+            if ($possibleType['name'] === $dataType || $possibleType['name'] === $alternativeName) {
+                return $context->getNavigator()->accept($data, $possibleType);
+            }
+        }
+    }
+
+    /**
+     *  ReflectionUnionType::getTypes() returns the types sorted according to these rules:
+     * - Classes, interfaces, traits, iterable (replaced by Traversable), ReflectionIntersectionType objects, parent and self:
+     *     these types will be returned first, in the order in which they were declared.
+     * - static and all built-in types (iterable replaced by array) will come next. They will always be returned in this order:
+     *     static, callable, array, string, int, float, bool (or false or true), null.
+     *
+     * For determining types of primitives, it is necessary to reorder primitives so that they are tested from lowest specificity to highest:
+     * i.e. null, true, false, int, float, bool, string
+     */
+    private function reorderTypes(array $type): array
```

Comment:
Let's move it into `TypedPropertiesDriver`. In that way it will be reorder once & cached, not for each deserialisation.

Reply:
Good idea.